### ISSUE TITLE
regexes: . different from [^\n]

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11913]
Migrated from [rt.perl.org#109206](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=109206) (status was 'resolved')

Searchable as RT109206$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



% perl \-wle '$\_ = "\\n"; print $\+\[0\] while /\[^\\n\]\*/g'
0
1

% perl \-wle '$\_ = "\\n"; print $\+\[0\] while /\.\*/g'    
0

I think this is a bug because in the absence of /s '\.' should match any
character except newline\, i\.e\. be equivalent to '\[^\\n\]'\. The two programs
should produce identical output\.

I also think the first result is correct because there are two zero\-length
matches in "\\n"\, one at the beginning of the string and one at the end\.
In conclusion&#8203;: it looks like /\.\*/g is broken\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl 5.12.1 - Thu Jun  3 20:09:15 CEST 2010
It is being executed now by  Perl 5.14.2 - Wed Sep 28 23:40:09 CEST 2011.

Site configuration information for perl 5.14.2:

Configured by mauke at Wed Sep 28 23:40:09 CEST 2011.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.38-gentoo-r6, archname=i686-linux
    uname='linux nora 2.6.38-gentoo-r6 #1 preempt sat aug 6 03:05:34 cest 2011 i686 amd athlon(tm) 64 processor 3200+ authenticamd gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -march=native -flto',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.1', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-fstack-protector -L/usr/local/lib -O2 -flto'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.12.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.12.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -march=native -flto -L/usr/local/lib -fstack-protector'

Locally applied patches:
    SAVEARGV0 - disable magic open in <ARGV>


@INC for perl 5.14.2:
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.2/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.2
    /home/mauke/usr/local/lib/perl5/5.14.2/i686-linux
    /home/mauke/usr/local/lib/perl5/5.14.2
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.1/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.1
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.0/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.0
    /home/mauke/usr/local/lib/perl5/site_perl
    .


Environment for perl 5.14.2:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/local/lib
    LOGDIR (unset)
    PATH=/home/mauke/usr/perlbrew/bin:/home/mauke/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.4.5:/opt/sun-jdk-1.4.2.13/bin:/opt/sun-jdk-1.4.2.13/jre/bin:/opt/sun-jdk-1.4.2.13/jre/javaws:/opt/dmd/bin:/usr/games/bin
    PERLBREW_HOME=/home/mauke/.perlbrew
    PERLBREW_PATH=/home/mauke/usr/perlbrew/bin
    PERLBREW_ROOT=/home/mauke/usr/perlbrew
    PERLBREW_VERSION=0.27
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @abigail

On Fri\, Jan 27\, 2012 at 06&#8203;:40&#8203;:09AM \-0800\, l\.mai@&#8203;web\.de wrote&#8203;:
> \# New Ticket Created by  l\.mai@&#8203;web\.de 
> \# Please include the string&#8203;:  \[perl \#109206\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=109206 >
> 
> 
> 
> This is a bug report for perl from l\.mai@&#8203;web\.de\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> % perl \-wle '$\_ = "\\n"; print $\+\[0\] while /\[^\\n\]\*/g'
> 0
> 1
> 
> % perl \-wle '$\_ = "\\n"; print $\+\[0\] while /\.\*/g'    
> 0
> 
> I think this is a bug because in the absence of /s '\.' should match any
> character except newline\, i\.e\. be equivalent to '\[^\\n\]'\. The two programs
> should produce identical output\.
> 
> I also think the first result is correct because there are two zero\-length
> matches in "\\n"\, one at the beginning of the string and one at the end\.
> In conclusion&#8203;: it looks like /\.\*/g is broken\.


I agree\. Note that if one makes the \* possessive\, it does give the 
same answer as when using \[^\\n\]&#8203;:

&nbsp;    $ perl \-wE '$\_ = "\\n"; say scalar \(\(\) = /\.\*/g\)'
&nbsp;    1
&nbsp;    $ perl \-wE '$\_ = "\\n"; say scalar \(\(\) = /\.\*\+/g\)'
&nbsp;    2
&nbsp;    $


Abigail

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 27 January 2012 15&#8203;:40\, l\.mai@&#8203;web\.de \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  l\.mai@&#8203;web\.de
> \# Please include the string&#8203;:  \[perl \#109206\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=109206 >
>
>
>
> This is a bug report for perl from l\.mai@&#8203;web\.de\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> % perl \-wle '$\_ = "\\n"; print $\+\[0\] while /\[^\\n\]\*/g'
> 0
> 1
>
> % perl \-wle '$\_ = "\\n"; print $\+\[0\] while /\.\*/g'
> 0
>
> I think this is a bug because in the absence of /s '\.' should match any
> character except newline\, i\.e\. be equivalent to '\[^\\n\]'\. The two programs
> should produce identical output\.
>
> I also think the first result is correct because there are two zero\-length
> matches in "\\n"\, one at the beginning of the string and one at the end\.
> In conclusion&#8203;: it looks like /\.\*/g is broken\.

This problem is caused by a broken optimisation\. The ANCH\_MBOL
optmisation\. Notice it the principle difference in these two outputs&#8203;:

$ perl  \-Mre=Debug\,DUMP \-wle '$\_ = "\\n"; print pos\($\_\)\,"&#8203;:"\,$\+\[0\] while /\.\*/g'
Compiling REx "\.\*"
Final program&#8203;:
&nbsp;   1&#8203;: STAR \(3\)
&nbsp;   2&#8203;:   REG\_ANY \(0\)
&nbsp;   3&#8203;: END \(0\)
anchored\(MBOL\) implicit minlen 0
0&#8203;:0
Freeing REx&#8203;: "\.\*"

$ perl  \-Mre=Debug\,DUMP \-wle '$\_ = "\\n"; print pos\($\_\)\,"&#8203;:"\,$\+\[0\] while
/\[^\\n\]\*/g'
Compiling REx "\[^\\n\]\*"
Final program&#8203;:
&nbsp;   1&#8203;: STAR \(13\)
&nbsp;   2&#8203;:   ANYOF\[\\0\-\\11\\13\-\\377\]\[\{unicode\_all\}\] \(0\)
&nbsp;  13&#8203;: END \(0\)
minlen 0
0&#8203;:0
1&#8203;:1
Freeing REx&#8203;: "\[^\\n\]\*"

It is enabled by this block of code in regcomp\. Notice the comment&#8203;:

 /\* turn \.\* into ^\.\* with an implied $\*=1 \*/

I have to admit I have not checked to see what the heck $\*=1 means\.

&nbsp;        else if \(\(\!sawopen || \!RExC\_sawback\) &&
&nbsp;            \(OP\(first\) == STAR &&
&nbsp;            PL\_regkind\[OP\(NEXTOPER\(first\)\)\] == REG\_ANY\) &&
&nbsp;            \!\(r\->extflags & RXf\_ANCH\) && \!\(RExC\_seen & REG\_SEEN\_EVAL\)\)
&nbsp;        \{
&nbsp;            /\* turn \.\* into ^\.\* with an implied $\*=1 \*/
&nbsp;            const int type =
&nbsp;                \(OP\(NEXTOPER\(first\)\) == REG\_ANY\)
&nbsp;                    ? RXf\_ANCH\_MBOL
&nbsp;                    : RXf\_ANCH\_SBOL;
&nbsp;            r\->extflags |= type;
&nbsp;            r\->intflags |= PREGf\_IMPLICIT;
&nbsp;            first = NEXTOPER\(first\);
&nbsp;            goto again;
&nbsp;        \}

The following patch disables the optimization&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/regcomp.c b/regcomp.c
index 668f8f7..12d0ac0 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -5235,7 +5235,7 @@ reStudy:
            first = NEXTOPER(first);
            goto again;
        }
-       else if ((!sawopen || !RExC_sawback) &&
+       else if (0 && (!sawopen || !RExC_sawback) &&
            (OP(first) == STAR &&
            PL_regkind[OP(NEXTOPER(first))] == REG_ANY) &&
            !(r->extflags & RXf_ANCH) && !(RExC_seen & REG_SEEN_EVAL))



```
</details>
Producing this output:
$ \./perl \-Ilib \-Mre=Debug\,DUMP \-wle '$\_ = "\\n"; print
pos\($\_\)\,"&#8203;:"\,$\+\[0\] while /\.\*/g'
Compiling REx "\.\*"
Final program&#8203;:
&nbsp;   1&#8203;: STAR \(3\)
&nbsp;   2&#8203;:   REG\_ANY \(0\)
&nbsp;   3&#8203;: END \(0\)
minlen 0
0&#8203;:0
1&#8203;:1
Freeing REx&#8203;: "\.\*"

I have not committed this patch as I dont know what effects it might
have\, however as it is a "conversion optimization" I would assume it
can be safely disabled until the underlying logic is fixed\. However I
will note that fixing it might be tricky\, the relevent code is spread
out over pp\_hot\.c and CALLREG\_INTUIT\_START\(\)\, and is particularly
hairy anyway\. It always makes me kinda cringe when I look at pp\_match\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 27 January 2012 16&#8203;:33\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> I have not committed this patch as I dont know what effects it might
> have\,

I decided to try out the smoke\-me thing\, and pushed it as

smoke\-me/disable\_anch\_mbol

Lets see what they say\.

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Jan 27 07&#8203;:33&#8203;:32 2012\, demerphq wrote&#8203;:
>  /\* turn \.\* into ^\.\* with an implied $\*=1 \*/
> 
> I have to admit I have not checked to see what the heck $\*=1 means\.

$\* doesn’t do anything anymore\, unless you are using Classic&#8203;::Perl\.

$\* = 1 puts /m on every match in 5\.8\, bugs aside\.

What the comment means exactly by implied $\*=1 I don’t know\.  Is it
referring to /^/ meaning /^/m in split?  But that couldn’t be right\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @mauke

On 2012\-01\-27 Father Chrysostomos via RT wrote&#8203;:

> On Fri Jan 27 07&#8203;:33&#8203;:32 2012\, demerphq wrote&#8203;:
> >  /\* turn \.\* into ^\.\* with an implied $\*=1 \*/
> > 
> > I have to admit I have not checked to see what the heck $\*=1 means\.
> 
> $\* doesn’t do anything anymore\, unless you are using Classic&#8203;::Perl\.
> 
> $\* = 1 puts /m on every match in 5\.8\, bugs aside\.
> 
> What the comment means exactly by implied $\*=1 I don’t know\.  Is it
> referring to /^/ meaning /^/m in split?  But that couldn’t be right\.

It means that a regexp that starts with \.\* is implicitly anchored
because if it doesn't match at offset 0\, it won't match at offsets 1\,
2\, 3 \.\.\. either\. /m is implied because \(since \.\* won't cross newlines\)
there can be multiple possible match locations if the string contains
\\n\. Which means you have to check every embedded \\n for a match\.

\(Conversely\, if /s is active\, leading \.\* should generate an implicit ^
with /m off \(a\.k\.a\. \\A\)\.\)

AFAICS this optimization is valid except when the target string ends
with a newline\. In that case \.\* could \(and should\) match\, but /^/m
won't\. That is\, "\\n" =~ /^/mg only matches once\.

So \.\.\. I guess the regex code should behave differently if the /^/m is
implicit and \\n is the last character in the target string?

\(And maybe there's a missed optimization opportunity here because I
don't see why this special case shouldn't trigger for \[^\\n\]\* at the
beginning of a pattern\.\)

**Comment by p5pRT** [comment]
### From @demerphq

On 27 January 2012 21&#8203;:51\, Lukas Mai \<l\.mai@&#8203;web\.de> wrote&#8203;:
> On 2012\-01\-27 Father Chrysostomos via RT wrote&#8203;:
>
>> On Fri Jan 27 07&#8203;:33&#8203;:32 2012\, demerphq wrote&#8203;:
>> >  /\* turn \.\* into ^\.\* with an implied $\*=1 \*/
>> >
>> > I have to admit I have not checked to see what the heck $\*=1 means\.
>>
>> $\* doesn’t do anything anymore\, unless you are using Classic&#8203;::Perl\.
>>
>> $\* = 1 puts /m on every match in 5\.8\, bugs aside\.

Ah\, thanks\. Pity the comment doesnt say "with an implied /m" instead\.

>> What the comment means exactly by implied $\*=1 I don’t know\.  Is it
>> referring to /^/ meaning /^/m in split?  But that couldn’t be right\.
>
> It means that a regexp that starts with \.\* is implicitly anchored
> because if it doesn't match at offset 0\, it won't match at offsets 1\,
> 2\, 3 \.\.\. either\. /m is implied because \(since \.\* won't cross newlines\)
> there can be multiple possible match locations if the string contains
> \\n\. Which means you have to check every embedded \\n for a match\.

Yes\, right\.

> \(Conversely\, if /s is active\, leading \.\* should generate an implicit ^
> with /m off \(a\.k\.a\. \\A\)\.\)
>
> AFAICS this optimization is valid except when the target string ends
> with a newline\. In that case \.\* could \(and should\) match\, but /^/m
> won't\. That is\, "\\n" =~ /^/mg only matches once\.

One might argue this is the bug\. It probably should match before and
after as well\.

>
> So \.\.\. I guess the regex code should behave differently if the /^/m is
> implicit and \\n is the last character in the target string?

Thing is the optimization is enabled before we ever see the string at
all\. It cannot depend on the contents of the string\.

So we either have to figure out how to make it match properly or
simply disable it\.

> \(And maybe there's a missed optimization opportunity here because I
> don't see why this special case shouldn't trigger for \[^\\n\]\* at the
> beginning of a pattern\.\)

Because it isnt easy to introspect the contents of a charclass\.


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @mauke

On 2012\-01\-28 demerphq wrote&#8203;:

> You say that you consider "\\n" to contain only one line\.
> 
> But what about "\\nfoo"\. Does it contain one or two lines? Do you
> expect ^ to match after the \\n in "\\nfoo"? If you do then do you not
> agree there is an inconsistency about it not matching after the \\n in
> "\\n"?

"\\nfoo" contains 1\.5 lines\, i\.e\. one complete \(but empty\) line and one
incomplete \(unterminated\) line\.
"\\nfoo" =~ /^foo/m should match\, yes\.
I don't think there's an inconsistency because \\n is only the beginning
of a line if more text follows\. That is\, my model of /^/m
is /\(?&#8203;:\\A|\(?\<=\\n\)\(?\!\\z\)\)/\.

> Because to me that is the exact same thing as expecting /\.\*/ to match
> at the end of the string in "\\n"\.
> 
> So really the bug here is in ^ not in \.\*

That doesn't match my intuitive understanding of "beginning of line"\.

**Comment by p5pRT** [comment]
### From @demerphq

On 28 January 2012 19&#8203;:30\, Lukas Mai \<l\.mai@&#8203;web\.de> wrote&#8203;:
> On 2012\-01\-28 demerphq wrote&#8203;:
>
>> You say that you consider "\\n" to contain only one line\.
>>
>> But what about "\\nfoo"\. Does it contain one or two lines? Do you
>> expect ^ to match after the \\n in "\\nfoo"? If you do then do you not
>> agree there is an inconsistency about it not matching after the \\n in
>> "\\n"?
>
> "\\nfoo" contains 1\.5 lines\, i\.e\. one complete \(but empty\) line and one
> incomplete \(unterminated\) line\.
> "\\nfoo" =~ /^foo/m should match\, yes\.
> I don't think there's an inconsistency because \\n is only the beginning
> of a line if more text follows\. That is\, my model of /^/m
> is /\(?&#8203;:\\A|\(?\<=\\n\)\(?\!\\z\)\)/\.

And the docs agree with you\, perlre says this&#8203;:

&nbsp;   You may\, however\, wish to treat a string as a multi\-line buffer\, such that
&nbsp;   the "^" will match after any newline within the string \(except if the newline
&nbsp;   is the last character in the string\)\, and "$" will match before any newline\.

Though I do wonder if the "except if the newline  is the last
character of the string" was a special case added later\.

>> Because to me that is the exact same thing as expecting /\.\*/ to match
>> at the end of the string in "\\n"\.
>>
>> So really the bug here is in ^ not in \.\*
>
> That doesn't match my intuitive understanding of "beginning of line"\.

It is sort of a lawyers point I guess\. To me the definition \(start of
string or immediately after a newline\) would match up with expecting
/\.\*/ to match twice against "\\n"\.

Anyway\, it sounds like the ANCH\_MBOL optimization is buggy\, so do we
turn it off or try to fix it somehow\.\.\.

cheers\,
Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @cpansprout

On Sat Jan 28 13&#8203;:27&#8203;:55 2012\, demerphq wrote&#8203;:
> On 28 January 2012 19&#8203;:30\, Lukas Mai \<l\.mai@&#8203;web\.de> wrote&#8203;:
> > On 2012\-01\-28 demerphq wrote&#8203;:
> >
> >> You say that you consider "\\n" to contain only one line\.
> >>
> >> But what about "\\nfoo"\. Does it contain one or two lines? Do you
> >> expect ^ to match after the \\n in "\\nfoo"? If you do then do you
> not
> >> agree there is an inconsistency about it not matching after the \\n
> in
> >> "\\n"?
> >
> > "\\nfoo" contains 1\.5 lines\, i\.e\. one complete \(but empty\) line and
> one
> > incomplete \(unterminated\) line\.
> > "\\nfoo" =~ /^foo/m should match\, yes\.
> > I don't think there's an inconsistency because \\n is only the
> beginning
> > of a line if more text follows\. That is\, my model of /^/m
> > is /\(?&#8203;:\\A|\(?\<=\\n\)\(?\!\\z\)\)/\.
> 
> And the docs agree with you\, perlre says this&#8203;:
> 
>    You may\, however\, wish to treat a string as a multi\-line buffer\,
> such that
>    the "^" will match after any newline within the string \(except if
> the newline
>    is the last character in the string\)\, and "$" will match before any
> newline\.
> 
> Though I do wonder if the "except if the newline  is the last
> character of the string" was a special case added later\.

Which is\, interestingly \(but irrelevantly\)\, the way JavaScript does it&#8203;:

$ perl \-MJE \-le 'print new JE\->eval\(q|/\\n^/m\.test\("\\n"\)|\)'true
true

\(Or enter javascript&#8203;:alert\(/\\n^/m\.test\("\\n"\)\) in a web browser\.\)

In JavaScript\, /^/m is equivalent to Perl’s
/\\A|\(?\<=\[\\cm\\cj\\x\{2028\}\\x\{2029\}\]\)/\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @demerphq

Thanks for your report\. I have fixed this in bleadperl with&#8203;:


commit 21eede782bed11b0263f9bff02b9ca7b7dfcd6eb
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Sun Jan 29 00&#8203;:06&#8203;:23 2012 \+0100

&nbsp;    Fix bug \#109206&#8203;: ANCH\_MBOL with while /\.\*/g
&nbsp;    
&nbsp;    We had a fencepost error when ANCH\_MBOL was enabled that meant we
&nbsp;    did not "see" matches at the end of string\. This fixes the problem
&nbsp;    and adds tests\.

Cheers\,
yves

**Comment by p5pRT** [comment]
@demerphq - Status changed from 'open' to 'resolved'
